rust_snuba/src/processors/eap_items.rs

```diff
@@ -0,0 +1,390 @@
+}
+
+seq_attrs! {
+#[derive(Debug, Default, Serialize)]
+pub(crate) struct AttributeMap {
+    attributes_bool: HashMap<String, bool>,
+    attributes_int: HashMap<String, i64>,
+    #(
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    attributes_string_~N: HashMap<String, String>,
+
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    attributes_float_~N: HashMap<String, f64>,
+    )*
+}
+}
+
+impl AttributeMap {
+    pub fn insert_str(&mut self, k: String, v: String) {
+        seq_attrs! {
+            let attr_str_buckets = [
+                #(
+                &mut self.attributes_string_~N,
+                )*
+            ];
+        };
+
+        attr_str_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_str_buckets.len()].insert(k, v);
+    }
+
+    pub fn insert_float(&mut self, k: String, v: f64) {
+        seq_attrs! {
+            let attr_num_buckets = [
+                #(
+                &mut self.attributes_float_~N,
+                )*
+            ];
+        }
+
+        attr_num_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_num_buckets.len()].insert(k, v);
+    }
+
+    pub fn insert_bool(&mut self, k: String, v: bool) {
+        // double write as float and bool
+        if v {
+            self.insert_float(k.clone(), 1.0);
+        } else {
+            self.insert_float(k.clone(), 0.0);
+        }
+        self.attributes_bool.insert(k, v);
+    }
+
+    pub fn insert_int(&mut self, k: String, v: i64) {
+        // double write as float and int
+        self.insert_float(k.clone(), v as f64);
+        self.attributes_int.insert(k, v);
+    }
+}
+
+#[derive(Debug, Default, Serialize, Ord, PartialOrd, Eq, PartialEq, Clone)]
+pub(crate) struct PrimaryKey {
+    pub organization_id: u64,
+    pub project_id: u64,
+    pub item_type: u8,
+    pub timestamp: u32,
+}
+
+#[derive(Debug, Default, Serialize)]
+struct EAPItem {
+    #[serde(flatten)]
+    primary_key: PrimaryKey,
+
+    trace_id: Uuid,
+    item_id: u128,
+    sampling_weight: u64,
+    retention_days: Option<u16>,
+
+    #[serde(flatten)]
+    attributes: AttributeMap,
+}
+
+fn fnv_1a(input: &[u8]) -> u32 {
+    const FNV_1A_PRIME: u32 = 16777619;
+    const FNV_1A_OFFSET_BASIS: u32 = 2166136261;
+
+    let mut res = FNV_1A_OFFSET_BASIS;
+    for byt in input {
+        res ^= *byt as u32;
+        res = res.wrapping_mul(FNV_1A_PRIME);
+    }
+
+    res
+}
+
+impl From<FromSpanMessage> for EAPItem {
+    fn from(from: FromSpanMessage) -> EAPItem {
+        let item_id = u128::from_str_radix(&from.span_id, 16).expect(&format!(
+            "Failed to parse span_id into u128. span_id: {}",
+            from.span_id
+        ));
+        let mut res = Self {
+            primary_key: PrimaryKey {
+                organization_id: from.organization_id,
+                project_id: from.project_id,
+                item_type: 1, // hardcoding "span" as type
+                timestamp: from.start_timestamp_ms as u32,
```

Comment:
This needs to be divided by 1000 since it's the start timestamp in milliseconds.